File: binarySearch_problems/UpperBound.js

/*

You are given an integer array of size n and you are given element 'x'. Return the index of the first value greater than 'x'.

*/

function findUpperBound(arr, x){
    let s = 0, e = arr.length-1;
    let ans = -1;

    while(s <= e){
        let mid = s + Math.floor((e - s)/2);
        if(arr[mid] <= x){
            // discard the left half
            s = mid + 1;
        }else{
            // in right half we will get the first probable ans
            ans = mid;
            e = mid - 1;
        }
    }
    return ans;
}

let arr = [2,2,3,3,3,4,6,6,8];
let res = findUpperBound(arr,4);
console.log(res);

File: 03.BinarySearch_problems/UpperBound.java
package BinarySearch_problems;

/**
 * UpperBound
 */
public class UpperBound {
    public static int findUpperBound(int[] arr, int x){
        int s = 0, e = arr.length - 1;
        int ans = -1;

        while(s <= e){
            int mid = (s + e)/2;
            // discard left half of array
            if(x >= arr[mid]){
                s = mid + 1;
            }else{
                // probable ans will available in right half
                ans = mid;
                e = mid - 1;
            }
        }
        return ans;
    }

    public static void main(String[] args) {
        int[] arr = {2,2,3,3,5,5,8,8};
        int ans = findUpperBound(arr, 5);
        System.out.println(ans);
    }
}